src/Rings/mpolyquo-localizations.jl

```diff
@@ -2717,3 +2717,90 @@ function dim(R::MPolyQuoLocRing)
+function _allunique(lst::Vector{T}) where {T<:MPolyLocRingElem}
+  return _allunique(numerator.(lst))
+end
+
+function _allunique(lst::Vector{T}) where {T<:MPolyQuoLocRingElem}
+  return _allunique(lifted_numerator.(lst))
+end
+
+_is_gen(x::MPolyLocRingElem) = is_one(denominator(x)) && _is_gen(numerator(x))
+_is_gen(x::MPolyQuoLocRingElem) = is_one(lifted_denominator(x)) && _is_gen(lifted_numerator(x))
+
+function _evaluate_plain(
+    F::MPolyAnyMap{<:MPolyRing, CT}, u
+  ) where {CT <: Union{<:MPolyLocRing, <:MPolyQuoLocRing}}
+  W = codomain(F)::MPolyLocRing
+  S = base_ring(W)
+  isdefined(F, :variable_indices) && return W(_build_poly(u, F.variable_indices, S))
+  return evaluate(u, F.img_gens)
+end
+
+function _evaluate_general(
+    F::MPolyAnyMap{<:MPolyRing, CT}, u
+  ) where {CT <: Union{<:MPolyQuoRing, <:MPolyLocRing, <:MPolyQuoLocRing}}
+  S = temp_ring(F)
+  if S !== nothing
+    if !isdefined(F, :variable_indices) || coefficient_ring(S) !== codomain(F)
+      return evaluate(map_coefficients(coefficient_map(F), u,
+                                       parent = S), F.img_gens)
+    else
+      tmp_poly = map_coefficients(coefficient_map(F), u, parent = S)
+      return _evaluate_with_build_ctx(tmp_poly, F.variable_indices, codomain(F))
+    end
+  else
+    if !isdefined(F, :variable_indices)
+      return evaluate(map_coefficients(coefficient_map(F), u), F.img_gens)
+    else
+      # For the case where we can recycle the method above, do so.
+      tmp_poly = map_coefficients(coefficient_map(F), u)
+      coefficient_ring(parent(tmp_poly)) === codomain(F) && return _evaluate_with_build_ctx(
+                                                                                            tmp_poly,
+                                                                                            F.variable_indices,
+                                                                                            codomain(F)
+                                                                                           )
+      # Otherwise default to the standard evaluation for the time being.
+      return evaluate(tmp_poly, F.img_gens)
+    end
+  end
+end
+
+function _evaluate_with_build_ctx(
+    p::MPolyRingElem, ind::Vector{Int}, 
+    Q::Union{<:MPolyQuoRing, <:MPolyLocRing, <:MPolyQuoLocRing}
+  )
+  @assert Q === coefficient_ring(parent(p))
+  cod_ring = base_ring(Q)
+  r = ngens(cod_ring)
+  kk = coefficient_ring(cod_ring)
+  ctx = MPolyBuildCtx(cod_ring)
+  for (q, e) in zip(coefficients(p), exponents(p))
+    ee = [0 for _ in 1:r]
+    for (i, k) in enumerate(e)
+      ee[ind[i]] = k
+    end
+    for (c, d) in zip(_coefficients(q), _exponents(q))
+      push_term!(ctx, kk(c), ee+d)
+    end
+  end
+  return Q(finish(ctx))
+end
+
+# The following methods are only safe, because they are called 
+# exclusively in a setup where variables map to variables 
+# and no denominators are introduced. 
+_coefficients(x::MPolyRingElem) = coefficients(x)
```

Comment:
Not sure, but maybe you want the following here?
```suggestion
_coefficients(x::MPolyRingElem) = AbstractAlgebra.coefficients(x)
```
And also below with `exponent_vectors`.

Reply:
I saw the `AA`-qualifier elsewhere, too. But I thought, it was redundant. Isn't it?

Reply:
No. It is the difference between painfully slow and fast when iterating over coefficients/exponents. There used to be an explanation in the documentation, but it was removed for reasons I don't know.

Reply:
See https://github.com/oscar-system/Oscar.jl/issues/4168